<h1>Frequently Asked Questions</h1>

<p>v0.3.17
Revised Thu Aug 27 13:53:48 EDT 2015</p>

<h2>What if none of these answers solve my problem?</h2>

<p>No problem. Just email me at qorts.contact at gmail dot com. If possible attach the output from QoRTs, and run QoRTs with the "--verbose" option to maximize the debugging information.</p>

<h2>Using QoRTs with different aligners:</h2>

<p>Different aligners often use different conventions, in particular with regards to the MAPQ field which is often used to differentiate multi-mapped and uniquely-mapped reads.</p>

<ul>
<li><strong>RNA-STAR</strong>: QoRTs is designed to work with RNA-STAR without any special options.</li>
<li><strong>TopHat1</strong>: Prior to version 2.0.0, TopHat used the same MAPQ convention as RNA-STAR now uses. Thus, no special options are needed.</li>
<li><strong>TopHat2</strong>: TopHat2 uses a MAPQ of 50 to mark reads that are uniquely aligned. Thus, when using TopHat2 aligned data, you must set the --minMAPQ parameter to 50.</li>
<li><strong>MapSplice</strong>: Under many test conditions MapSplice produces malformed SAM files that violate the <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">current SAM format specification</a>. As such, QoRTs may not be able to properly assess MapSplice output, particularly for paired-ended data.</li>
<li><strong>GSNAP</strong>: I have been unable to find documentation describing exactly what the MAPQ field means in GSNAP output. However, setting the --minMAPQ parameter to 20 seems to work.</li>
<li><strong>BowTie, BowTie2, BWA, NOVOALIGN, BLAT, BLAST, or ELEND</strong>: The use of older aligners or aligners not specifically designed for use with RNA-Seq data is NOT recommended or supported. You MIGHT be able to get these alignments to work using --minMAPQ 0.</li>
</ul>

<h2>How do I install QoRTs?</h2>

<p>The easiest way to install QoRTs is using the R command:</p>

<pre><code>&gt; install.packages("http://hartleys.github.io/QoRTs/QoRTs_LATEST.tar.gz",
                   repos=NULL, 
                   type="source");
</code></pre>

<h2>Manual Installation</h2>

<p>Alternatively you can download QoRTs directly from the <a href="https://github.com/hartleys/QoRTs/releases/latest">release page</a>, and install using the command:</p>

<pre><code>&gt; install.packages("QoRTs_0.3.17.tar.gz",repos=NULL, type="source");
</code></pre>

<p>Note: If you run into an error that looks like this:</p>

<pre><code>ERROR: cannot extract package from '0.3.17.tar.gz'
</code></pre>

<p>Then you may be attempting to install the entire github repository archive, rather than the actual R package archive. Download the QoRTs_0.3.9.zip file from <a href="https://github.com/hartleys/QoRTs/releases/latest">the most recent release</a>, unzip it, and use the file "QoRTs_0.3.9.tar.gz" to install the R package. Do NOT attempt to install the full repository source in R!</p>

<p>The java jar file does not need to be installed. It can be downloaded from the <a href="https://github.com/hartleys/QoRTs/releases/latest">release page</a>.
Just execute it using the java command:</p>

<pre><code>java -jar /path/to/jarfile/QoRTs.jar QC input.bam anno.gtf.gz /output/dir/
</code></pre>

<h2>Permission errors on installation:</h2>

<p>If, during package installation, you encounter an error like:</p>

<pre><code>&gt; Warning in install.packages("QoRTs", lib = "~/.R/library") :
&gt;  'lib = "~/.R/library"' is not writable
</code></pre>

<p>Then you may not have write permissions for your package directory. You can install QoRTs to a different directory using the R command:</p>

<pre><code>&gt; install.packages("http://hartleys.github.io/QoRTs/QoRTs_LATEST.tar.gz",
                   repos=NULL, 
                   type="source", 
                   lib="path/to/package/dir");
</code></pre>

<p>Make sure the "path/to/package/dir" directory exists and is writable. Then you can load this local copy of QoRTs using the R command:</p>

<pre><code>&gt; library("QoRTs",lib.loc="path/to/package/dir/")
</code></pre>

<h2>Non-standard Phred scores</h2>

<p>If you see an warning that looks like this:</p>

<pre><code>WARNING WARNING WARNING: 
SAM format check:
    Phred Qual &gt; 41!
    You will need to set either --adjustPhredScores or --maxPhredScores
    in order to compute Phred quality metrics! QoRTs WILL throw an error
    if quality metrics are attempted!
</code></pre>

<p>Followed by an error that looks like this:</p>

<pre><code>ERROR! ArrayIndexOutOfBoundsException caught while attempting to store quality score metrics.
       Most likely cause: quality score found above 41!
</code></pre>

<p>Then you may have a problem with your Phred score encoding. This can be caused
by one of two possibilities:</p>

<p>(1) Certain newer Illumina platforms (reportedly) produce Phred scores in excess of 41. 
While not explicitly forbidden by the Phred encoding specification (such as it is), 
this is very uncommon at the current time. However, if a Phred score in excess of
41 is encountered then QoRTs will throw a descriptive error like this:</p>

<p>In this case, you will need to use the "--maxPhredScore" parameter, set to the maximum expected
Phred score. According to one (secondhand) report, Illumina stated that the absolute maximum 
(as of August 2015) was 45.</p>

<p>(2) Certain older Illumina platforms (~1.3-1.7) used Phred+64 encoding for their phred scores. 
While technically SAM files should always be converted to Phred+33, QoRTs does offer 
optional support for reading such malformed input files. Use the "--adjustPhredScore 31"
option to automatically subtract 31 from all phred scores, thus converting Phred+64 to Phred+33.
Note that negative Phred scores are NOT supported by QoRTs, and they WILL throw an error.
See the <a href="https://en.wikipedia.org/wiki/FASTQ_format#Encoding">related wikipedia article</a> for more information. </p>

<h2>How do I cite QoRTs?</h2>

<p>If you use QoRTs and find it helpful, you can cite it in your publications as:</p>

<p>Hartley SW, Mullikin JC. <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4506620/"><strong>QoRTs: a comprehensive toolset for quality control and data processing of RNA-Seq experiments.</strong></a> <em>BMC Bioinformatics.</em>](http://www.biomedcentral.com/bmcbioinformatics) doi: 10.1186/s12859-015-0670-5</p>
