<!DOCTYPE html>

<html>
		<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="chrome=1">
		<title>QoRTs: Quality of RNA-Seq Toolset</title>

		<link rel="stylesheet" href="stylesheets/styles.css">
		<link rel="stylesheet" href="stylesheets/pygment_trac.css">
		<script src="javascripts/scale.fix.js"></script>
		<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

		<!--[if lt IE 9]>
		<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
		<![endif]-->
</head>

<body>
<div class="wrapper">


<header>
			<h1><a href="index.html">QoRTs</a></h1>
			<p>Quality of RNA-seq Toolset</p>
			<p class="view"><a href="https://github.com/hartleys/QoRTs">View the Project on GitHub <small>Stephen Hartley</small></a></p>
			<ul>
				<li><a href="https://github.com/hartleys/QoRTs/releases/latest">Latest <strong>Release</strong></a></li>
				<li><a href="https://github.com/hartleys/QoRTs/zipball/master">Download <strong>ZIP File</strong></a></li>
				<li><a href="https://github.com/hartleys/QoRTs/tarball/master">Download <strong>TAR Ball</strong></a></li>
				<li><a href="https://github.com/hartleys/QoRTs">View On <strong>GitHub</strong></a></li>
			</ul>
</header>


<section><h1>Frequently Asked Questions</h1>

<p>v0.3.14
Revised Fri Aug 21 11:07:55 EDT 2015</p>

</section> <h2>What if none of these answers solve my problem?</h2> <section>

No problem. Just email me at qorts.contact at gmail dot com. If possible attach the output from QoRTs, and run QoRTs with the "--verbose" option to maximize the debugging information.

</section> <h2>Using QoRTs with different aligners:</h2> <section>


Different aligners often use different conventions, in particular with regards to the MAPQ field which is often used to differentiate multi-mapped and uniquely-mapped reads.

<ul>
<li> <b>RNA-STAR</b>: QoRTs is designed to work with RNA-STAR without any special options. </li>
<li> <b>TopHat1</b>: Prior to version 2.0.0, TopHat used the same MAPQ convention as RNA-STAR now uses. Thus, no special options are needed. </li>
<li> <b>TopHat2</b>: TopHat2 uses a MAPQ of 50 to mark reads that are uniquely aligned. Thus, when using TopHat2 aligned data, you must set the --minMAPQ parameter to 50. </li>
<li> <b>MapSplice</b>: Under many test conditions MapSplice produces malformed SAM files that violate the current SAM format specification. As such, QoRTs may not be able to properly assess MapSplice output, particularly for paired-ended data. </li>
<li> <b>GSNAP</b>: I have been unable to find documentation describing exactly what the MAPQ field means in GSNAP output. However, setting the --minMAPQ parameter to 20 seems to work. </li>
<li> <b>BowTie, BowTie2, BWA, NOVOALIGN, BLAT, BLAST, or ELEND</b>: The use of older aligners or aligners not specifically designed for use with RNA-Seq data is NOT recommended or supported. You MIGHT be able to get these alignments to work using --minMAPQ 0. </li>
</ul>


</section> <h2>Problems with Installation:</h2> <section>

<p>The R package can be installed in R using the command:</p>

<pre><code>&gt; install.packages("QoRTs_0.3.14.tar.gz", repos = NULL, type="source")
</code></pre>

<p>or using the command-line:</p>

<pre><code>R CMD INSTALL QoRTs_0.3.14.tar.gz
</code></pre>

Note: If you run into an error that looks like this:
<pre><code>
ERROR: cannot extract package from ‘QoRTs_0.3.9.tar.gz’
</code></pre>
then you may be attempting to install the github repository tar.gz archive, rather than the actual R package archive. Download the QoRTs_0.3.9.zip file from <a href="https://github.com/hartleys/QoRTs/releases/latest">the most recent release</a>, unzip it, and use the file "QoRTs_(versionNumber)tar.gz" to install the R package.

<p>The java jar file does not need to be installed.
Just execute it using the java command:</p>

<pre><code>java -jar /path/to/jarfile/QoRTs.jar QC input.bam anno.gtf.gz /output/dir/
</code></pre>

</section> <h2>EXAMPLE DATA:</h2> <section>

<p>The example QC output can be found on the github main page, and installed
with the command:</p>

<pre><code>install.packages("QoRTsExampleData_0.3.14.tar.gz", repos = NULL, type="source")
</code></pre>

<p>The original bam files are too large to upload to github.
These files, along with a comprehensive walkthrough demonstrating
how all analysis can be run on them, can be found <a href="
https://dl.dropboxusercontent.com/u/103621176/qorts/exData/QoRTsFullExampleData.zip">here</a>.
(File is ~1.5gb)</p>

</section> <h2>How do I cite QoRTs?</h2> <section>
If you use QoRTs and find it helpful, you can cite it in your publications as:<br>

Hartley SW, Mullikin JC. <b><a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4506620/">QoRTs: a comprehensive toolset for quality control and data processing of RNA-Seq experiments.</a></b> <i><a href="http://www.biomedcentral.com/bmcbioinformatics">BMC Bioinformatics.</a></i> doi: 10.1186/s12859-015-0670-5

</section> <h2>How do I access the internal command-line help?</h2> <section>

<p>Additional options and syntax information for the main QC java utility
can be found using the command:</p>

<pre><code>java -jar /path/to/jarfile/QoRTs.jar QC --man
</code></pre>

<p>Options and information about other sub-utilities within the java package
can be found using the command:</p>

<pre><code>java -jar /path/to/jarfile/QoRTs.jar --man
</code></pre>

<p>And for each sub-utility:</p>

<pre><code>java -jar /path/to/jarfile/QoRTs.jar utilname --man
</code></pre>

<p>For help with individual R functions in the R utility, use the R command:</p>

<pre><code>&gt; help(functionname);
</code></pre>

<p>For a full listing of all help topics for the R utility, use the R command: </p>

<pre><code>&gt; help(package="QoRTs");
</code></pre>

</section>

</div>

		<footer> <small>
		<p>Project maintained by <a href="https://github.com/hartleys/QoRTs">Stephen Hartley</a></p>
		<p>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a> (modified)</p>
		</small> </footer>

</body></html>
